Source/WebCore/platform/cocoa/ParentalControlsURLFilter.mm

```diff
@@ -0,0 +1,99 @@
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#import "config.h"
+#import "ParentalControlsURLFilter.h"
+
+#if HAVE(WEBCONTENTRESTRICTIONS)
+
+#import <wtf/CompletionHandler.h>
+#import <wtf/MainThread.h>
+#import <wtf/URL.h>
+#import <wtf/cocoa/VectorCocoa.h>
+
+#import <pal/cocoa/WebContentRestrictionsSoftLink.h>
+
+namespace WebCore {
+
+WTF_MAKE_TZONE_ALLOCATED_IMPL(ParentalControlsURLFilter);
+
+static bool wcrBrowserEngineClientEnabled()
+{
+    return [PAL::getWCRBrowserEngineClientClass() shouldEvaluateURLs];
+}
+
+static RefPtr<ParentalControlsURLFilter>& globalDefaultParentalControlsURLFilter()
+{
+    ASSERT(isMainThread());
+
+    static NeverDestroyed<RefPtr<ParentalControlsURLFilter>> filter;
+    return filter;
+}
+
+Ref<ParentalControlsURLFilter> ParentalControlsURLFilter::defaultParentalControlsURLFilter()
+{
+    ASSERT(isMainThread());
+
+    auto& filter = globalDefaultParentalControlsURLFilter();
+    if (!filter)
+        filter = adoptRef(*new ParentalControlsURLFilter);
+
+    return *filter;
+}
+
+ParentalControlsURLFilter::ParentalControlsURLFilter()
+{
+    if (wcrBrowserEngineClientEnabled())
+        m_wcrBrowserEngineClient = adoptNS([PAL::allocWCRBrowserEngineClientInstance() init]);
+}
+
+bool ParentalControlsURLFilter::enabled()
+{
+    return !!m_wcrBrowserEngineClient;
+}
+
+void ParentalControlsURLFilter::isURLAllowed(const URL& url, CompletionHandler<void(bool, Vector<uint8_t>&&)>&& completionHandler)
+{
+    if (!m_wcrBrowserEngineClient)
+        return completionHandler(true, { });
+
+    [m_wcrBrowserEngineClient evaluateURL:url withCompletion:makeBlockPtr([completionHandler = WTFMove(completionHandler)](BOOL shouldBlock, NSData *replacementData) mutable {
+        completionHandler(!shouldBlock, makeVector(replacementData));
```

Comment:
Are we sure this always runs on the main thread? We should at least assert it.

Reply:
Also, it seems it would be more efficient to call `span(replacementData)` and pass a span to the completion handler instead of a vector. We never end up needing a vector, just a span so there is no reason to copy the bytes to a vector here I think.

Reply:
Changed `CompletionHandler` to take `NSData*` to avoid copy.